File: lib/permission-helper/lifecycle.js
// lifecycle.js — Listener state machine for the permission helper.
//
// State diagram (from tech design):
//
//   WAITING (idle ≤ 10 min, heartbeats present)
//     → /apply received     → APPLYING
//     → /reject received    → REJECTED → exit(1)
//     → heartbeat absent    → DISCONNECTED → exit(4)
//     → idle ≥ 10 min       → TIMED_OUT → exit(2)
//     → SIGINT/SIGTERM      → TERMINATED → exit(130|143)
//
//   APPLYING (apply-script running)
//     → apply-script exits 0       → DONE
//     → apply-script exits 1|2     → ERROR_AWAITING_USER
//     → apply-script exits 3       → ERROR_INTERNAL → exit(3)
//     → SIGINT/SIGTERM             → TERMINATED → exit(130|143)
//
//   DONE (apply complete, post-completion window 30s)
//     → window expires             → exit(0)
//     → page closed                → exit(0)
//     → SIGINT/SIGTERM             → exit(130|143)
//
//   ERROR_AWAITING_USER (partial/total failure, user can retry)
//     → /retry received            → APPLYING (with reduced spec)
//     → /reject received           → exit(3)
//     → heartbeat absent ≥ 15s     → exit(3)
//     → idle ≥ 10 min              → exit(2)
//     → SIGINT/SIGTERM             → exit(130|143)

'use strict';

const STATES = Object.freeze({
  WAITING: 'WAITING',
  APPLYING: 'APPLYING',
  DONE: 'DONE',
  REJECTED: 'REJECTED',
  TIMED_OUT: 'TIMED_OUT',
  DISCONNECTED: 'DISCONNECTED',
  TERMINATED: 'TERMINATED',
  ERROR_AWAITING_USER: 'ERROR_AWAITING_USER',
  ERROR_INTERNAL: 'ERROR_INTERNAL',
});

const EXIT_CODES = Object.freeze({
  APPLIED: 0,
  REJECTED: 1,
  TIMED_OUT: 2,
  APPLY_FAILED: 3,
  PAGE_CLOSED: 4,
  VALIDATION_ERROR: 5,
  SIGINT: 130,
  SIGTERM: 143,
});

const DEFAULTS = Object.freeze({
  idleTimeoutMs: 10 * 60 * 1000,           // 10 min
  heartbeatGapMs: 15 * 1000,                // 15 sec
  heartbeatExpectedIntervalMs: 5 * 1000,    // page sends every 5s
  postCompletionWindowMs: 30 * 1000,        // 30 sec after DONE
  applyScriptHardTimeoutMs: 5 * 60 * 1000,  // 5 min — kill apply-script if no exit by then
});

class Lifecycle {
  /**
   * @param {object}   options
   * @param {function} options.onTerminal  Callback invoked when a terminal state is reached.
   *                                        Receives ({state, exitCode, statusReport}).
   * @param {object}   [options.timing]    Override default timeouts (used in tests).
   */
  constructor({ onTerminal, timing = {} }) {
    if (typeof onTerminal !== 'function') {
      throw new TypeError('Lifecycle requires onTerminal callback');
    }
    this.onTerminal = onTerminal;
    this.timing = { ...DEFAULTS, ...timing };
    this.state = STATES.WAITING;
    this.lastHeartbeat = Date.now();
    this.statusReport = {
      outcome: null,
      applied_operations: [],
      failed_operations: [],
      verified_post_state: [],
      user_edits: null,
      error_detail: null,
    };
    this._timers = {};
    this._terminated = false;
    this._startIdleTimer();
    this._startHeartbeatWatcher();
  }

  // ---- Public transitions (called by the listener on incoming events) ----

  onApply(spec) {
    if (this.state !== STATES.WAITING && this.state !== STATES.ERROR_AWAITING_USER) {
      return false;
    }
    this._setState(STATES.APPLYING);
    this._clearTimers();
    this._startApplyHardTimeout();
    this.statusReport.user_edits = this._captureUserEdits(spec);
    return true;
  }

  onReject() {
    if (this.state === STATES.WAITING) {
      this._terminate(STATES.REJECTED, EXIT_CODES.REJECTED, { outcome: 'rejected' });
      return true;
    }
    if (this.state === STATES.ERROR_AWAITING_USER) {
      this._terminate(STATES.REJECTED, EXIT_CODES.APPLY_FAILED, { outcome: 'apply_error', error_detail: 'user closed without retry after partial failure' });
      return true;
    }
    return false;
  }

  onRetry(spec) {
    if (this.state !== STATES.ERROR_AWAITING_USER) return false;
    this._setState(STATES.APPLYING);
    this._clearTimers();
    this._startApplyHardTimeout();
    this.statusReport.user_edits = this._captureUserEdits(spec);
    return true;
  }

  onHeartbeat() {
    this.lastHeartbeat = Date.now();
    return true;
  }

  onApplyScriptExit(exitCode, applyResult) {
    if (this.state !== STATES.APPLYING) return false;
    this._clearTimers();

    // applyResult is the parsed `done` line from apply-script stdout.
    if (applyResult) {
      this.statusReport.applied_operations = applyResult.successful || [];
      this.statusReport.failed_operations = applyResult.failed || [];
      this.statusReport.verified_post_state = applyResult.verified_post_state || [];
    }

    if (exitCode === 0) {
      this._setState(STATES.DONE);
      this.statusReport.outcome = 'applied';
      this._startPostCompletionWindow();
    } else if (exitCode === 1 || exitCode === 2) {
      this._setState(STATES.ERROR_AWAITING_USER);
      this.statusReport.outcome = exitCode === 1 ? 'partial_failure' : 'apply_error';
      this._startIdleTimer();
      this._startHeartbeatWatcher();
    } else {
      // exitCode === 3 or unexpected
      this.statusReport.outcome = 'apply_error';
      this.statusReport.error_detail = `apply-script exited ${exitCode}`;
      this._terminate(STATES.ERROR_INTERNAL, EXIT_CODES.APPLY_FAILED, this.statusReport);
    }
    return true;
  }

  onSignal(sig) {
    const exitCode = sig === 'SIGINT' ? EXIT_CODES.SIGINT : EXIT_CODES.SIGTERM;
    this.statusReport.outcome = 'terminated';
    this._terminate(STATES.TERMINATED, exitCode, this.statusReport);
  }

  onPageClose() {
    // Heartbeat-gap detection delegates here.
    if (this.state === STATES.WAITING || this.state === STATES.ERROR_AWAITING_USER) {
      const code = this.state === STATES.ERROR_AWAITING_USER ? EXIT_CODES.APPLY_FAILED : EXIT_CODES.PAGE_CLOSED;
      this.statusReport.outcome = this.state === STATES.ERROR_AWAITING_USER ? 'apply_error' : 'page_closed';
      this._terminate(STATES.DISCONNECTED, code, this.statusReport);
    } else if (this.state === STATES.DONE) {
      // User saw the success page and closed it. Exit cleanly.
      this._terminate(STATES.DONE, EXIT_CODES.APPLIED, this.statusReport);
    }
    // APPLYING: do NOT cancel — let the apply-script run to completion.
    // Heartbeat absence during APPLYING is logged but does not terminate.
  }

  // ---- Internal ----

  _setState(newState) {
    this.state = newState;
  }

  _terminate(state, exitCode, report) {
    if (this._terminated) return;
    this._terminated = true;
    this._setState(state);
    this._clearTimers();
    Object.assign(this.statusReport, report);
    this.onTerminal({ state, exitCode, statusReport: this.statusReport });
  }

  _captureUserEdits(submittedSpec) {
    // The listener passes the originally-rendered spec and the page-submitted spec.
    // For simplicity, the listener stores edits as a diff that's surfaced here.
    // If submittedSpec is just the spec, we record presence; the listener supplies the diff.
    return submittedSpec ? { spec_at_apply: submittedSpec } : null;
  }

  _startIdleTimer() {
    this._clearTimer('idle');
    this._timers.idle = setTimeout(() => {
      this.statusReport.outcome = 'timed_out';
      this._terminate(STATES.TIMED_OUT, EXIT_CODES.TIMED_OUT, this.statusReport);
    }, this.timing.idleTimeoutMs);
  }

  _startHeartbeatWatcher() {
    this._clearTimer('heartbeat');
    const checkInterval = Math.floor(this.timing.heartbeatExpectedIntervalMs / 2);
    this._timers.heartbeat = setInterval(() => {
      if (this._terminated) return;
      const since = Date.now() - this.lastHeartbeat;
      if (since > this.timing.heartbeatGapMs) {
        this.onPageClose();
      }
    }, checkInterval);
  }

  _startPostCompletionWindow() {
    this._clearTimer('postCompletion');
    this._timers.postCompletion = setTimeout(() => {
      this._terminate(STATES.DONE, EXIT_CODES.APPLIED, this.statusReport);
    }, this.timing.postCompletionWindowMs);
  }

  _startApplyHardTimeout() {
    this._clearTimer('applyHard');
    this._timers.applyHard = setTimeout(() => {
      this.statusReport.outcome = 'apply_error';
      this.statusReport.error_detail = `apply-script did not exit within ${this.timing.applyScriptHardTimeoutMs}ms; assumed crashed`;
      this._terminate(STATES.ERROR_INTERNAL, EXIT_CODES.APPLY_FAILED, this.statusReport);
    }, this.timing.applyScriptHardTimeoutMs);
  }

  _clearTimer(name) {
    if (this._timers[name]) {
      clearTimeout(this._timers[name]);
      clearInterval(this._timers[name]);
      delete this._timers[name];
    }
  }

  _clearTimers() {
    for (const name of Object.keys(this._timers)) this._clearTimer(name);
  }
}

module.exports = { Lifecycle, STATES, EXIT_CODES, DEFAULTS };

File: lib/permission-helper-go/internal/listener/lifecycle.go
// Package listener provides the ephemeral one-shot HTTP listener and
// the lifecycle state machine for the permission helper.
//
// State diagram (mirrored from the Node helper's lifecycle.js):
//
//   WAITING          → APPLYING (on /apply) | REJECTED (on /reject)
//                    | DISCONNECTED (heartbeat gap >15s) | TIMED_OUT (idle 10min)
//                    | TERMINATED (signal)
//
//   APPLYING         → DONE (apply exit 0) | ERROR_AWAITING_USER (apply exit 1|2)
//                    | ERROR_INTERNAL (apply exit 3) | TERMINATED (signal)
//
//   DONE             → exit(0) after post-completion window or page close
//
//   ERROR_AWAITING_USER → APPLYING (on /retry) | exit(3) (on /reject or page close)
//                       | TIMED_OUT | TERMINATED
package listener

import (
	"sync"
	"time"
)

type State string

const (
	StateWaiting           State = "WAITING"
	StateApplying          State = "APPLYING"
	StateDone              State = "DONE"
	StateRejected          State = "REJECTED"
	StateTimedOut          State = "TIMED_OUT"
	StateDisconnected      State = "DISCONNECTED"
	StateTerminated        State = "TERMINATED"
	StateErrorAwaitingUser State = "ERROR_AWAITING_USER"
	StateErrorInternal     State = "ERROR_INTERNAL"
)

// Exit codes — keep in sync with the Node helper's EXIT_CODES.
const (
	ExitApplied         = 0
	ExitRejected        = 1
	ExitTimedOut        = 2
	ExitApplyFailed     = 3
	ExitPageClosed      = 4
	ExitValidationError = 5
	ExitSIGINT          = 130
	ExitSIGTERM         = 143
)

// Default timing (mirrored from the Node helper's DEFAULTS).
type Timing struct {
	IdleTimeoutMs              int
	HeartbeatGapMs             int
	HeartbeatExpectedIntervalMs int
	PostCompletionWindowMs     int
	ApplyScriptHardTimeoutMs   int
}

func DefaultTiming() Timing {
	return Timing{
		IdleTimeoutMs:               10 * 60 * 1000,
		HeartbeatGapMs:              15 * 1000,
		HeartbeatExpectedIntervalMs: 5 * 1000,
		PostCompletionWindowMs:      30 * 1000,
		ApplyScriptHardTimeoutMs:    5 * 60 * 1000,
	}
}

// StatusReport is the final JSON the binary writes to stdout when it exits.
// Same shape as the Node helper's emitFinal().
type StatusReport struct {
	Outcome           string        `json:"outcome"`
	AppliedOperations []int         `json:"applied_operations"`
	FailedOperations  []FailedOp    `json:"failed_operations"`
	VerifiedPostState []VerifiedRow `json:"verified_post_state"`
	UserEdits         interface{}   `json:"user_edits"`
	ErrorDetail       string        `json:"error_detail,omitempty"`
}

type FailedOp struct {
	OpIndex int        `json:"op_index"`
	Error   ErrorBlock `json:"error"`
}

type ErrorBlock struct {
	Code    string `json:"code"`
	Message string `json:"message,omitempty"`
}

type VerifiedRow struct {
	OpIndex    int                  `json:"op_index"`
	Resource   string               `json:"resource"`
	Recipients []map[string]string  `json:"recipients"`
}

// Terminal is the data passed to the OnTerminal callback.
type Terminal struct {
	State        State
	ExitCode     int
	StatusReport StatusReport
}

// Lifecycle is the state machine. Created with NewLifecycle, fed events
// via the OnX methods, and produces a single Terminal call when one of
// the terminal conditions is reached.
type Lifecycle struct {
	mu             sync.Mutex
	state          State
	timing         Timing
	lastHeartbeat  time.Time
	statusReport   StatusReport
	terminated     bool
	timers         lifecycleTimers
	onTerminal     func(Terminal)
}

type lifecycleTimers struct {
	idle           *time.Timer
	heartbeat      *time.Ticker
	stopHeartbeat  chan struct{}
	postCompletion *time.Timer
	applyHard      *time.Timer
}

// NewLifecycle constructs a Lifecycle in StateWaiting and starts the
// idle timer + heartbeat watcher.
func NewLifecycle(onTerminal func(Terminal), timing Timing) *Lifecycle {
	if onTerminal == nil {
		panic("listener: NewLifecycle requires onTerminal callback")
	}
	lc := &Lifecycle{
		state:         StateWaiting,
		timing:        timing,
		lastHeartbeat: time.Now(),
		onTerminal:    onTerminal,
	}
	lc.statusReport.AppliedOperations = []int{}
	lc.statusReport.FailedOperations = []FailedOp{}
	lc.statusReport.VerifiedPostState = []VerifiedRow{}
	lc.startIdleTimer()
	lc.startHeartbeatWatcher()
	return lc
}

// State returns the current state. Useful for tests and HTTP gating.
func (lc *Lifecycle) State() State {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	return lc.state
}

// OnApply transitions to APPLYING. Returns false if not in a state that
// can begin applying.
func (lc *Lifecycle) OnApply() bool {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	if lc.state != StateWaiting && lc.state != StateErrorAwaitingUser {
		return false
	}
	lc.state = StateApplying
	lc.clearTimers()
	lc.startApplyHardTimeout()
	return true
}

// OnReject terminates with REJECTED.
func (lc *Lifecycle) OnReject() bool {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	if lc.state == StateWaiting {
		lc.statusReport.Outcome = "rejected"
		lc.terminate(StateRejected, ExitRejected)
		return true
	}
	if lc.state == StateErrorAwaitingUser {
		lc.statusReport.Outcome = "apply_error"
		lc.statusReport.ErrorDetail = "user closed without retry after partial failure"
		lc.terminate(StateRejected, ExitApplyFailed)
		return true
	}
	return false
}

// OnRetry transitions ERROR_AWAITING_USER back to APPLYING.
func (lc *Lifecycle) OnRetry() bool {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	if lc.state != StateErrorAwaitingUser {
		return false
	}
	lc.state = StateApplying
	lc.clearTimers()
	lc.startApplyHardTimeout()
	return true
}

// OnHeartbeat resets the page-disconnect watchdog.
func (lc *Lifecycle) OnHeartbeat() {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	lc.lastHeartbeat = time.Now()
}

// OnApplyScriptExit reports the apply-script's exit code and final report.
func (lc *Lifecycle) OnApplyScriptExit(exitCode int, applyResult *StatusReport) bool {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	if lc.state != StateApplying {
		return false
	}
	lc.clearTimers()

	if applyResult != nil {
		lc.statusReport.AppliedOperations = applyResult.AppliedOperations
		lc.statusReport.FailedOperations = applyResult.FailedOperations
		lc.statusReport.VerifiedPostState = applyResult.VerifiedPostState
	}

	switch exitCode {
	case 0:
		lc.state = StateDone
		lc.statusReport.Outcome = "applied"
		lc.startPostCompletionWindow()
	case 1:
		lc.state = StateErrorAwaitingUser
		lc.statusReport.Outcome = "partial_failure"
		lc.startIdleTimer()
		lc.startHeartbeatWatcher()
	case 2:
		lc.state = StateErrorAwaitingUser
		lc.statusReport.Outcome = "apply_error"
		lc.startIdleTimer()
		lc.startHeartbeatWatcher()
	default:
		lc.statusReport.Outcome = "apply_error"
		lc.statusReport.ErrorDetail = "apply-script exited with unexpected code"
		lc.terminate(StateErrorInternal, ExitApplyFailed)
	}
	return true
}

// OnSignal handles SIGINT/SIGTERM.
func (lc *Lifecycle) OnSignal(sig string) {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	exitCode := ExitSIGTERM
	if sig == "SIGINT" {
		exitCode = ExitSIGINT
	}
	lc.statusReport.Outcome = "terminated"
	lc.terminate(StateTerminated, exitCode)
}

// OnPageClose handles a heartbeat gap or beforeunload beacon.
func (lc *Lifecycle) OnPageClose() {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	switch lc.state {
	case StateWaiting:
		lc.statusReport.Outcome = "page_closed"
		lc.terminate(StateDisconnected, ExitPageClosed)
	case StateErrorAwaitingUser:
		lc.statusReport.Outcome = "apply_error"
		lc.terminate(StateDisconnected, ExitApplyFailed)
	case StateDone:
		// User saw success and closed the page. Exit cleanly.
		lc.terminate(StateDone, ExitApplied)
	}
	// In APPLYING: do NOT cancel — let the apply-script run to completion.
}

// HeartbeatGapExceeded checks whether the page is silent for too long.
// Called periodically by the heartbeat watcher goroutine.
func (lc *Lifecycle) HeartbeatGapExceeded() bool {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	gap := time.Since(lc.lastHeartbeat)
	return gap > time.Duration(lc.timing.HeartbeatGapMs)*time.Millisecond
}

// terminate is internal — caller must hold lc.mu.
func (lc *Lifecycle) terminate(state State, exitCode int) {
	if lc.terminated {
		return
	}
	lc.terminated = true
	lc.state = state
	lc.clearTimers()
	cb := lc.onTerminal
	report := lc.statusReport
	// Release the lock before invoking the callback to avoid deadlocks
	// if the callback tries to call back into Lifecycle methods.
	go cb(Terminal{State: state, ExitCode: exitCode, StatusReport: report})
}

// startIdleTimer — caller must hold lc.mu.
func (lc *Lifecycle) startIdleTimer() {
	if lc.timers.idle != nil {
		lc.timers.idle.Stop()
	}
	lc.timers.idle = time.AfterFunc(time.Duration(lc.timing.IdleTimeoutMs)*time.Millisecond, func() {
		lc.mu.Lock()
		defer lc.mu.Unlock()
		if lc.terminated {
			return
		}
		lc.statusReport.Outcome = "timed_out"
		lc.terminate(StateTimedOut, ExitTimedOut)
	})
}

// startHeartbeatWatcher — caller must hold lc.mu.
func (lc *Lifecycle) startHeartbeatWatcher() {
	if lc.timers.stopHeartbeat != nil {
		close(lc.timers.stopHeartbeat)
	}
	stop := make(chan struct{})
	lc.timers.stopHeartbeat = stop
	checkInterval := time.Duration(lc.timing.HeartbeatExpectedIntervalMs/2) * time.Millisecond
	lc.timers.heartbeat = time.NewTicker(checkInterval)
	go func() {
		ticker := lc.timers.heartbeat
		defer ticker.Stop()
		for {
			select {
			case <-stop:
				return
			case <-ticker.C:
				if lc.HeartbeatGapExceeded() {
					lc.OnPageClose()
					return
				}
			}
		}
	}()
}

// startPostCompletionWindow — caller must hold lc.mu.
func (lc *Lifecycle) startPostCompletionWindow() {
	if lc.timers.postCompletion != nil {
		lc.timers.postCompletion.Stop()
	}
	lc.timers.postCompletion = time.AfterFunc(time.Duration(lc.timing.PostCompletionWindowMs)*time.Millisecond, func() {
		lc.mu.Lock()
		defer lc.mu.Unlock()
		if lc.terminated {
			return
		}
		lc.terminate(StateDone, ExitApplied)
	})
}

// startApplyHardTimeout — caller must hold lc.mu.
func (lc *Lifecycle) startApplyHardTimeout() {
	if lc.timers.applyHard != nil {
		lc.timers.applyHard.Stop()
	}
	lc.timers.applyHard = time.AfterFunc(time.Duration(lc.timing.ApplyScriptHardTimeoutMs)*time.Millisecond, func() {
		lc.mu.Lock()
		defer lc.mu.Unlock()
		if lc.terminated {
			return
		}
		lc.statusReport.Outcome = "apply_error"
		lc.statusReport.ErrorDetail = "apply-script hard timeout exceeded"
		lc.terminate(StateErrorInternal, ExitApplyFailed)
	})
}

// clearTimers — caller must hold lc.mu.
func (lc *Lifecycle) clearTimers() {
	if lc.timers.idle != nil {
		lc.timers.idle.Stop()
		lc.timers.idle = nil
	}
	if lc.timers.postCompletion != nil {
		lc.timers.postCompletion.Stop()
		lc.timers.postCompletion = nil
	}
	if lc.timers.applyHard != nil {
		lc.timers.applyHard.Stop()
		lc.timers.applyHard = nil
	}
	if lc.timers.stopHeartbeat != nil {
		close(lc.timers.stopHeartbeat)
		lc.timers.stopHeartbeat = nil
	}
}
